### PR TITLE
Add Python 3 support

### PR DESCRIPTION
**Pull request by bertinatto**
Do we want to support `py3[0-2]`?

If we do want to support this version range we have to go over all string literals and use `six.u` instead of the `u` prefix literal. I'm avoiding using `unicode_literals`.

### CODE REVIEW

**Comment by chekunkov** [comment]
> Do we want to support py3[0-2]?
don't think so, Python>=3.3 is fine

>  I'm avoiding using unicode_literals.
why?

**Comment by bertinatto** [comment]
@chekunkov, I see 2 main reasons: first because we lose the option to specify a native string literal (or at least it is most more cumbersome [1]). It may be useful in the future.

Also, `unicode_literals` converts docstrings to unicode, crushing Pydoc in some cases (again, we may want to use it in the future :) ).

Armin Ronacher (flask) started a nice discussion about `unicode_literals` in this thread [2].

[1] http://python-future.org/unicode_literals.html#drawbacks
[2] https://github.com/PythonCharmers/python-future/issues/22

**Comment by chekunkov** [comment]
I see, thanks @bertinatto

File: hubstorage/batchuploader.py [hunk 1]
```diff
@@ -250,17 +252,21 @@ def __str__(self):
 
 
 def _encode_identity(iter):
-    data = StringIO()
+    data = BytesIO()
     for item in iter:
```
Comment (chekunkov) [inline hubstorage/batchuploader.py]: sorry for nitpicking, but can we replace `iter` with something like `iterable` here - I don't like when builtins are shadowed

File: hubstorage/resourcetype.py [hunk 2]
```diff
@@ -26,7 +28,9 @@ def _iter_lines(self, _path, **kwargs):
 
         r = self.client.request(**kwargs)
 
-        return r.iter_lines()
+        lines = r.iter_lines()
+        return (l.decode('utf8') if isinstance(l, six.binary_type)
+                else l for l in lines)
```
Comment (chekunkov) [inline hubstorage/resourcetype.py]: it looks like this method returned generator that produced strings, and now it produces unicode. why is that?

File: tests/test_retry.py [hunk 3]
```diff
@@ -76,7 +76,7 @@ def request_callback(request):
             if attempts_count[0] <= 2:
                 raise ConnectionError("Connection aborted.", BadStatusLine("''"))
             if attempts_count[0] == 3:
-                return (429, {}, {})
+                return (429, {}, u'')
```
Comment (chekunkov) [inline tests/test_retry.py]: why it changed?

File: hubstorage/batchuploader.py (hunk 1 above)
Comment (bertinatto) [inline hubstorage/batchuploader.py]: Sure. Done.

I handled it in another commit because it was already there before this PR.

File: hubstorage/resourcetype.py (hunk 2 above)
Comment (bertinatto) [inline hubstorage/resourcetype.py]: I found it more appropriate to handle them internally as unicode strings. Also, it looks more convenient in certain cases.

For example, in `serialization.py` [1], `json.loads()` expects a unicode string in Python 3+:
```

In [4]: json.loads(b'{"a":"a"}')
(...)
raise JSONDecodeError("Unexpected UTF-8 BOM (decode using utf-8-sig)",
TypeError: the JSON object must be str, not 'bytes'
```
On the other hand, in Python 2 `json.loads()` accepts both (unicode and bytes), but always generates objects with unicode strings:

```
In [4]: json.loads(b'{"a":"a"}')
Out[4]: {u'a': u'a'}

In [5]: json.loads(u'{"a":"a"}')
Out[5]: {u'a': u'a'}
```

[1] https://github.com/scrapinghub/python-hubstorage/blob/6098e59db1ad9bd04a60b650c4797d2cd7437f10/hubstorage/serialization.py#L17

File: tests/test_retry.py (hunk 3 above)
Comment (bertinatto) [inline tests/test_retry.py]: It was supposed to return a string rather than a dict, like here [1]. Python 3 wouldn't accept any type other than `str`.

[1] https://github.com/scrapinghub/python-hubstorage/blob/6098e59db1ad9bd04a60b650c4797d2cd7437f10/hubstorage/serialization.py#L17

File: hubstorage/batchuploader.py [hunk 4]
```diff
@@ -3,12 +3,14 @@
 import random
 import logging
 import warnings
+import six
```
Comment (chekunkov) [inline hubstorage/batchuploader.py]: please add six to requirements and setup.py

Comment (bertinatto) [inline hubstorage/batchuploader.py]: Thanks for catching this.

Added `six` and rebased against `master`. #64 allowed to catch 2 more issues.

File: hubstorage/resourcetype.py (hunk 2 above)
Comment (immerrr) [inline hubstorage/resourcetype.py]: if you really want to return unicode, you should do `response.iter_lines(decode_unicode=True)`

File: hubstorage/utils.py [hunk 5]
```diff
@@ -35,12 +36,10 @@ def urlpathjoin(*parts):
             continue
         elif isinstance(p, tuple):
             p = urlpathjoin(*p)
-        elif isinstance(p, unicode):
-            p = p.encode('utf8')
-        elif not isinstance(p, str):
-            p = str(p)
+        elif not isinstance(p, six.text_type):
+            p = six.u(str(p))
```
Comment (immerrr) [inline hubstorage/utils.py]: this should be simply `six.text_type(p)`

File: hubstorage/resourcetype.py (hunk 2 above)
Comment (chekunkov) [inline hubstorage/resourcetype.py]: The reason I'm asking is because it is not internal change - it changes output for public `iter_json` methods:

in master

```
In [5]: list(job.items.iter_json())
Out[5]:
['{"_key":"2222886/1/103/0","uuid":0}',
...
```

in py3 branch (Python 2.7.10)

```
In [6]: list(job.items.iter_json())
Out[6]:
[u'{"_key":"2222886/1/103/0","uuid":0}',
...
```

There could be cases where users rely on `iter_json` output to be strings in Python 2. Don't you think we need to keep backwards compatibility in Py2? 

I'm okay to keep it simple in Python3 - no need to output bytes, method says it iterates through data as strings.

Comment (bertinatto) [inline hubstorage/resourcetype.py]: @chekunkov, I couldn't reproduce the issue you had.

In `py3` branch:

```
In [20]: help()

Welcome to Python 2.7!  This is the online help utility.

(...)

In [21]: next(job.items.iter_json())

Out[21]: '{"_key":"1886/661/8/0","category":"\xe5\xae\xa0\xe7\x89\xa9\xe7\x94\xa8\xe5\x93\x81 > \xe7\x8b\x97\xe7\x94\xa8\xe5\x93\x81 > \xe7\x8b\x97\xe7\xb2\xae","_type":"DiscoveryItem","url_product_page":"http://www.amazon.cn/dp/B00CAO9M80","name":"american distribution and mfg co 00106 Nutri Life, 5 LB, Gold Plus Formula, Chicken, Duck, Fish, and Lamb Dog...","response_url":"http://www.amazon.cn/s/ref=sr_nr_n_13?fst=as%3Aoff&rh=n%3A118863071%2Cn%3A149410071%2Cn%3A149411071%2Ck%3Adog&keywords=dog&ie=UTF8&qid=1419962169&rnid=118864071","url":"http://www.amazon.cn/s?ie=UTF8&page=1&rh=n%3A149411071%2Ck%3Adog","country":"China","price_now":"135.21","image_urls":["http://ec4.images-amazon.com/images/I/51EKVXspZTL._AA160_.jpg"],"scrape_datetime":"2016-03-17T17:34:34Z","download_screenshot":false,"identifier_sku":"B00CAO9M80","page_number":1,"products_on_page":24,"country_code":"CN","page_item_number":1,"images":[],"shelf_type":"Regular","download_html":false}'
```

In `master` branch:

```
In [8]: help()

Welcome to Python 2.7!  This is the online help utility.

In [9]: next(job.items.iter_json())

Out[9]: '{"_key":"1886/661/8/0","category":"\xe5\xae\xa0\xe7\x89\xa9\xe7\x94\xa8\xe5\x93\x81 > \xe7\x8b\x97\xe7\x94\xa8\xe5\x93\x81 > \xe7\x8b\x97\xe7\xb2\xae","_type":"DiscoveryItem","url_product_page":"http://www.amazon.cn/dp/B00CAO9M80","name":"american distribution and mfg co 00106 Nutri Life, 5 LB, Gold Plus Formula, Chicken, Duck, Fish, and Lamb Dog...","response_url":"http://www.amazon.cn/s/ref=sr_nr_n_13?fst=as%3Aoff&rh=n%3A118863071%2Cn%3A149410071%2Cn%3A149411071%2Ck%3Adog&keywords=dog&ie=UTF8&qid=1419962169&rnid=118864071","url":"http://www.amazon.cn/s?ie=UTF8&page=1&rh=n%3A149411071%2Ck%3Adog","country":"China","price_now":"135.21","image_urls":["http://ec4.images-amazon.com/images/I/51EKVXspZTL._AA160_.jpg"],"scrape_datetime":"2016-03-17T17:34:34Z","download_screenshot":false,"identifier_sku":"B00CAO9M80","page_number":1,"products_on_page":24,"country_code":"CN","page_item_number":1,"images":[],"shelf_type":"Regular","download_html":false}'
```

I also created a test to make sure it always return a `unicode` string (`str` in py3):

screenshot from 2016-04-21 17 55 58

Comment (bertinatto) [inline hubstorage/resourcetype.py]: About `iter_lines(decode_unicode=True)`, the response coming from Hubstorage has no encoding sett, so `requests` library doesn't decode the data ¹.

¹https://github.com/kennethreitz/requests/blob/e65d6fc6378826e70304b7db660ab37d306dff54/requests/utils.py#L362

Comment (chekunkov) [inline hubstorage/resourcetype.py]: > the response coming from Hubstorage has no encoding set

how so? `Content-Type: application/x-jsonlines; charset=UTF-8`

> I couldn't reproduce the issue you had.

I deleted the branch and pulled it again - the issue persists, please send me the whole output (py3 test, change branch, master output) in private message

Comment (bertinatto) [inline hubstorage/resourcetype.py]: > I deleted the branch and pulled it again - the issue persists

I apologize for that. I was using an experimental Hubstorage branch that was returning back arbitrary results. I was able to reproduce the issue.

I agree we should keep backwards compatibility in py2. I've pushed a changed [1] to handle that.

I've also added a unit test [2] to make sure the output is compatible with the respective python version.

> how so? Content-Type: application/x-jsonlines; charset=UTF-8

If I set `decode_unicode=True` in py2 I get a generator that produces unicode strings, as expected. However, in py3 the generator sometimes produces bytes strings and sometimes unicode strings.

Inspecting `r.encoding` here [3] I noticed that it's occasionally set as `None`. I don't know if this is an issue with `requests` or Hubstorage (or if it's an issue at all), but I can't rely only on `decode_unicode=True` because it doesn't guarantee [4] the decoding.

[1] https://github.com/scrapinghub/python-hubstorage/blob/cece2665987ebc7acdf1471ea3a4cdc7cc9e7428/hubstorage/resourcetype.py#L32-L34
[2] https://github.com/scrapinghub/python-hubstorage/pull/62/commits/cece2665987ebc7acdf1471ea3a4cdc7cc9e7428
[3] https://github.com/scrapinghub/python-hubstorage/blob/31525d31c536eacfc19f416404e64ab21598f1d2/hubstorage/resourcetype.py#L27
[4] http://docs.python-requests.org/en/master/api/#requests.Response.iter_content

File: tests/test_batchuploader.py [hunk 6]
```diff
@@ -47,19 +48,19 @@ def test_writer_maxitemsize(self):
             ValueTooLarge,
             'Value exceeds max encoded size of 1048576 bytes:'
             ' \'{"b+\\.\\.\\.\'',
-            w.write, {'b'*(m/2): 'x'*(m/2)})
+            w.write, {'b'*int(m/2): 'x'*int(m/2)})
```
Comment (immerrr) [inline tests/test_batchuploader.py]: This is better expressed as `floordiv`, `m // 2`.

File: hubstorage/serialization.py [hunk 7]
```diff
@@ -29,5 +30,4 @@ def jsondefault(o):
         millis = (u + (s + d * ADAYINSECONDS) * 1e6) / 1000
```
Comment (immerrr) [inline hubstorage/serialization.py]: A floordiv here would be better, too. Py3 does float division by default.

File: hubstorage/resourcetype.py [hunk 8]
```diff
@@ -26,7 +28,10 @@ def _iter_lines(self, _path, **kwargs):
 
         r = self.client.request(**kwargs)
 
-        return r.iter_lines()
+        lines = r.iter_lines()
+        if six.PY3:
+            return (l.decode('utf8') for l in lines)
```
Comment (immerrr) [inline hubstorage/resourcetype.py]: `r.encoding or 'utf8'` maybe?

**Comment by immerrr** [comment]
Added couple nitpicks, but all in all seems nice. Do we want to enforce absolute imports with `from __future__ import absolute_import`?

File: hubstorage/resourcetype.py [hunk 9]
```diff
@@ -194,7 +199,7 @@ def save(self):
             if not self.ignore_fields:
                 self.apipost(jl=self._data, is_idempotent=True)
             else:
-                self.apipost(jl=dict((k, v) for k, v in self._data.iteritems()
+                self.apipost(jl=dict((k, v) for k, v in six.iteritems(self._data)
```
Comment (dangra) [inline hubstorage/resourcetype.py]: what about updating the python idioms: `{k: v for k, v in ... }`

**Comment by chekunkov** [comment]
> Do we want to enforce absolute imports with from __future__ import absolute_import

no objections

File: tests/test_batchuploader.py (hunk 6 above)
Comment (chekunkov) [inline tests/test_batchuploader.py]: +1 

also PEP8 suggests spaces between operands, i.e. `{'b' * (m // 2): 'x' * (m // 2)}`

Comment (immerrr) [inline tests/test_batchuploader.py]: to be fair, they are not necessary: https://www.python.org/dev/peps/pep-0008/#other-recommendations

pep8 requires one space around comparisons, boolean operators and assignments. other operators are required to have no more than one space and the same amount on both sides. also, quite a shocker for me from `No` section:
```
No:

x = x * 2 - 1
hypot2 = x * x + y * y
c = (a + b) * (a - b)
```

Comment (chekunkov) [inline tests/test_batchuploader.py]: interesting, I was taught to do this differently :) and linters usually show an error for such formatting. I don't have a strong opinion here, "Use your own judgment" can be applied.

**Comment by bertinatto** [comment]
> Do we want to enforce absolute imports with from future import absolute_import

May I question the use of `from future import absolute_import`? Do you think it's really necessary?

I don't see a lot of advantages using it, since Python 3 tests will catch the use of non-absolute imports. Also, I see it's not enforced in other py3-compatible libraries.

What do you think?

**Comment by immerrr** [comment]
> Do you think it [absolute import]'s really necessary?

No, this is why it was a question :) Ok, it would only affect intra-hubstorage import, we have control over it in tests, let's not do that.
